### PR TITLE
feat(asset/http): POST /api/v1/utxos bulk lookup + drop redundant tx fetch on GetUTXO

### PR DESCRIPTION
**Pull request by icellan**
Addresses #945 — adds a bulk UTXO spend-status endpoint and fixes the underlying slowness that made the single-UTXO endpoint not benefit from client-side parallelism.

## Summary

- **`stores/utxo/{aerospike,sql}.GetSpend`**: tolerate `spend.UTXOHash == nil`. The store already locates the record by primary key `(txid, vout)` and knows the canonical hash; the caller-supplied hash was a defensive check. Every existing caller still passes a hash, so the invariant is preserved for them — only the new HTTP fast paths opt out. The SQL path also stops dereferencing nil (was a latent panic).
- **`services/asset/httpimpl/GetUTXO.go`**: stop calling `GetTransaction` + `NewTxFromBytes` + `UTXOHashFromOutput`. Each `GET /api/v1/utxo/:hash?vout=N` is now one Aerospike record read instead of *(blob-fetch on cache miss) + (full tx parse) + (Aerospike Get)*. This is the latency the discussion was about.
- **`services/asset/httpimpl/GetUTXOs.go` (new)**: `POST /api/v1/utxos`. Binary request body of 36-byte `(txid || voutLE)` records; binary response of fixed 48-byte records `(statusLE || lockTimeLE || vinLE || spendingTxID)` in input order, zero-padded when unspent so clients can index by `i*48`. Fan-out via `errgroup` capped at 1024, preallocated response with per-slot writes (no mutex). Body size capped by a new `asset_httpBodyLimit` setting (default 4MB, ≈116k records).

## Behaviour change to flag

The old single-endpoint distinguished \"transaction not found\" from \"output index out of range\" in the 404 message. Both now collapse to a uniform `NOT_FOUND (3): UTXO not found`. The HTTP status is unchanged. Clients keying off the message string need to adjust; clients keying off status code do not.

## Test plan

- [x] `go test -race ./services/asset/httpimpl/...` (includes new `TestGetUTXOs` and updated `TestGetUTXO`)
- [x] `go test -race ./stores/utxo/sql/...` (includes new `TestGetSpendNilUTXOHash`; one pre-existing unrelated failure `TestMinedThenSpendAllPrunes_*` documented separately)
- [x] `go test ./stores/utxo/aerospike/...` (full integration suite, exercises the nil-tolerance change)
- [x] `go vet`, `staticcheck`, `golangci-lint` clean on touched packages
- [ ] Reviewer to spot-check the binary record layout against client expectations (see record layout in `GetUTXOs.go` doc comment)

### CODE REVIEW

File: services/asset/httpimpl/GetUTXO.go [hunk 1]
```diff
@@ -118,35 +116,16 @@ func (h *HTTP) GetUTXO(mode ReadMode) func(c echo.Context) error {
...
+		// raw transaction just to recompute a hash the store already has.
+		// On Status_NOT_FOUND below we treat output-out-of-range and
+		// txid-unknown uniformly as 404.
 		utxoResponse, err := h.repository.GetUtxo(ctx, &utxo.Spend{
 			TxID:         txHash,
 			Vout:         voutUint32,
-			UTXOHash:     utxoHash,
+			UTXOHash:     nil,
 			SpendingData: nil,
 		})
```
[the remark this answers is not in the document]
Comment (icellan) [reply services/asset/httpimpl/GetUTXO.go:130]: Fixed in a3e578afb. The log line at repository.go:917 now uses `TxID:Vout` (both always populated) instead of `UTXOHash.String()`. Added `TestGetUtxoNilUTXOHash` that exercises the real `Repository` against a sqlitememory store — the handler-layer tests used `asset/repository.Mock` which doesn't traverse this code path, which is why the bug wasn't caught initially.

File: services/asset/httpimpl/GetUTXOs.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+			vout := binary.LittleEndian.Uint32(reqBytes[recOffset+32 : recOffset+36])
+
+			g.Go(func() error {
+				resp, err := h.repository.GetUtxo(gCtx, &utxo.Spend{
+					TxID:         &txHash,
+					Vout:         vout,
+					UTXOHash:     nil,
+					SpendingData: nil,
+				})
```
[the remark this answers is not in the document]
Comment (icellan) [reply services/asset/httpimpl/GetUTXOs.go]: Same fix as the comment above: repository.go now logs `TxID:Vout` (a3e578afb). Both `POST /api/v1/utxos` and the simplified `GET /api/v1/utxo` paths pass `UTXOHash: nil` deliberately — the store locates the record by primary key. Added `TestGetUtxoNilUTXOHash` exercising the real `Repository` so the nil-hash path is covered in CI going forward.

**Review by oskarszoon — APPROVED** [review]
Approve with discussion items. 573/573 `-race` clean. Settings loader correctly wired this time (`settings.go:215` reads `asset_httpBodyLimit` — #933 / #643 P1.1 pattern doesn't repeat).

### Worth confirming before merge

1. **Aerospike `GetSpend` never populates `LockTime`** (`stores/utxo/aerospike/get.go:272-275`). SQL path does via `coinbaseSpendingHeight`. Binary/HEX response's lockTime is always 0 on Aerospike backend. Pre-existing, but bulk endpoint makes it machine-visible. Intentional → document in swagger; oversight → one-line fix.

2. **Endpoint anonymous + unrated.** `apiGroup` only wires Recover + BanList + CORS + Gzip + securityHeaders. PR #643 (auth + rate limit) still CHANGES_REQUESTED. Single 4MB POST = ~116k Aerospike `Get`s at 1024 concurrency through the shared `uaerospike.Client` semaphore (~256) that validator + block validation + propagation also use. Per-request cost is ~116k× the old single-UTXO endpoint. Worth a tracking issue for gating under `asset_httpHeavyRateLimit` once #643 lands.

3. **Dropped `UTXOHash` check** was the only run-time detector for writer-side `vout % batchSize` slot-indexing regressions in Aerospike. Future writer bug would silently return wrong data. Add writer-side regression test (`GetSpend(vout=X, UTXOHash=nil)` after `Spend(vout=Y)` with batchSize > 1) or hash-mismatch metric.

4. **Doc gaps:**
   - `asset_httpBodyLimit` missing from `docs/references/settings/services/asset_settings.md` — same shape as #643 v3.
   - `asset_reference.md:771-774` doesn't call out the message-collapse behaviour change ("tx not found" + "vout out of range" → "NOT_FOUND (3): UTXO not found"). Scripted clients break silently.

### Nits

- `GetUTXO.go:132` loose `strings.Contains(err.Error(), "not found")` while `GetUTXOs.go:149` uses `errors.Is(err, errors.ErrNotFound)` — tighten singular.
- `GetUTXOs_test.go` mixes `assert` with project `require`-only convention.
- `GetUTXO.go:37` docstring calls `:hash` "UTXO hash"; it's the txid.
- No 413 body-cap test — `GetMockHTTP` bypasses BodyLimit middleware.
- `c.Request().RemoteAddr` in tracing log at `GetUTXOs.go:92` — wrong behind a reverse proxy, consistent with pre-existing `GetUTXO.go:91`.

errgroup fan-out at 1024 safe (per-slot writes, `i := i` capture, client disconnect cancels `gCtx`). Binary layout 8+4+4+32=48 verified. `UTXOHash == nil` correct in both stores; SQL latent nil-deref fixed at `sql.go:3320`. `a3e578afb` nil-safe logging verified end-to-end. Mode dispatch closure-bound at route registration — no header-tricking surface.

**Comment by ordishs** [comment]
## Review

Clean design overall — the binary layout is well-documented, the store nil-`UTXOHash` tolerance and the repo logging fix are correct and necessary, and test coverage is broad (order preservation, mixed statuses, nil-hash assertions, HEX/JSON modes, the real-repo nil-log test). One **critical, remotely-triggerable crash** blocks merge.

---

### 🔴 CRITICAL — Unauthenticated DoS: out-of-range `vout` panics the aerospike store and crashes the asset server

Both new paths now pass a **client-controlled `vout` straight to `GetSpend` with no bounds check**. The aerospike store indexes the per-batch UTXO list without validating the offset:

`stores/utxo/aerospike/get.go:186`
```go
b, ok := utxos[spend.Vout%uint32(s.utxoBatchSize)].([]byte)
```

- `CalculateKeySource(hash, vout, 128)` returns `num = vout/128`, so **every `vout` in `0..127` maps to the same batch-0 record**, which exists for every tx (`util/uaerospike/client.go:505`).
- The `Utxos` bin is `utxos[start:end]`, sized to the **actual output count, not padded** (`stores/utxo/aerospike/create.go:574-577`). A 2-output tx stores a 2-element list.
- So `utxos[vout%128]` with e.g. `vout=127` on a 2-element list → **`index out of range` panic**.

This indexing is pre-existing but was previously **unreachable** with bad vouts: the only HTTP caller (`GetUTXO`) validated `voutUint32 >= uint32(len(tx.Outputs))` → 404 **before** hitting the store. This PR removes that guard and adds a second unvalidated path.

Impact differs by endpoint — and the bulk one is fatal:

| Path | Panic goroutine | `middleware.Recover()` covers it? | Result |
|---|---|---|---|
| `GET /api/v1/utxo/:hash?vout=N` | echo request goroutine | ✅ yes (`http.go:129`) | recovered → **500** (regression from the old clean 404) |
| `POST /api/v1/utxos` | **errgroup child goroutine** (`GetUTXOs.go:524`) | ❌ **no** — `recover()` is goroutine-local; errgroup does not recover | **unrecovered panic → asset process crashes** |

Full chain for the bulk path has **no recover anywhere**: `GetUTXOs` goroutine → `repository.GetUtxo` (`repository.go:911`) → aerospike `GetSpend` (panics). Echo's recover middleware only protects the request goroutine, not the ones `errgroup` spawns.

**Reproduction (unauthenticated, one request):**
```
POST /api/v1/utxos
body = <32-byte txid of ANY tx with <128 outputs> || <uint32 LE vout in (numOutputs, 127]>
```
The coinbase of any block (1 output, well-known txid) with `vout=1` suffices.

**Required fix (root cause, protects all callers, minimal):** bounds-check the offset in `aerospike/get.go` and treat overflow as not-found, mirroring the existing `ErrKeyNotFound` branch:
```go
idx := spend.Vout % uint32(s.utxoBatchSize)
if int(idx) >= len(utxos) {
    return &utxo.SpendResponse{Status: int(utxo.Status_NOT_FOUND)}, nil
}
b, ok := utxos[idx].([]byte)
```
SQL is already safe (`WHERE ... AND o.idx = $2` → `ErrNoRows` → `Status_NOT_FOUND`, `sql.go:3303-3310`), so the gap is **aerospike-only (production)**.

**Strongly recommend additionally** isolating panics in the `errgroup` body so any future store-level panic degrades to 500 instead of taking down the process — a public fan-out endpoint should never let a per-record panic escape the goroutine.

**Test gap that hid this:** every `GetUTXOs`/`GetUTXO` test mocks the repository, so "Vout out of range" returns a *mocked* `Status_NOT_FOUND` and never exercises the real store indexing. Per the repo's testing rule (use sqlitememory, don't mock the store), an out-of-range-vout test against a real store would have caught the aerospike-vs-SQL divergence. Please add one.

---

### 🟡 Minor

- **`413` can be masked as `500`.** `GetUTXOs` reads the body via `io.ReadAll` (`GetUTXOs.go:493`). Echo's `BodyLimit` returns 413 up-front only via the `Content-Length` check; a client streaming without an accurate `Content-Length` trips the limit inside `Read`, so `io.ReadAll` returns `ErrStatusRequestEntityTooLarge` and the handler wraps it as `500`. Common clients send `Content-Length` so the documented 413 holds for them — consider detecting `echo.ErrStatusRequestEntityTooLarge` and surfacing 413.
- **Behavior change (already flagged by author):** "tx not found" vs "output index out of range" now collapse to `NOT_FOUND (3): UTXO not found`. Documented, status code unchanged — just confirm no internal client greps the old message strings.
- **Error counter only on success.** `prometheusAssetHTTPGetUTXOs` is incremented only on the 200 path; 400/413/500 record nothing. Mirrors existing `GetUTXO`, so the bulk endpoint's failure rate will be invisible.

### 🟢 Nits
- `i := i` (`GetUTXOs.go:516`) is dead since Go 1.22; `go.mod` is `go 1.26.0`.
- `writeUTXOsRecord` doc says `[...4 vin][32 txid]`; PR body / `asset_reference.md` say "spendingVin"/"spendingTxID" — same thing, align wording.

---

### Verdict

**Request changes.** Endpoint design, docs, and test breadth are strong. But removing the `vout < len(outputs)` guard exposes a pre-existing unchecked slice index in the aerospike store, and routing it through `errgroup` goroutines turns it into a one-request remote crash of the asset server.

Before merge:
1. Bounds-check the offset in `aerospike/get.go` (return `Status_NOT_FOUND`).
2. Add panic isolation around the `errgroup` per-record body.
3. Add a real-store (sqlitememory + aerospike) out-of-range-vout test.
4. *(Optional)* map `ErrStatusRequestEntityTooLarge` from `io.ReadAll` to 413.

**Comment by icellan** [comment]
Thanks for the careful review @ordishs — the panic chain is exactly as you described and the test gap is a fair critique. Pushed `287eb7085` addressing every item.

## Critical fixes

1. **Aerospike bounds-check.** `stores/utxo/aerospike/get.go:186` now guards `idx >= len(utxos)` and returns `Status_NOT_FOUND`, mirroring the existing `ErrKeyNotFound` branch. SQL was already safe.
2. **Errgroup panic isolation.** Each `GetUTXOs` goroutine body now runs under a deferred `recover()` that converts any panic into a 500 errgroup error and logs the stack trace. A per-record panic can no longer escape the request goroutine, regardless of which store backend is in use.
3. **Real-store regression tests.** Added:
   - `stores/utxo/sql/sql_test.go::TestGetSpendNilUTXOHash/out-of-range_vout_returns_NOT_FOUND` (sqlitememory, runs locally).
   - `stores/utxo/aerospike/aerospike_server_test.go::TestAerospike/aerospike_get_spend_out_of_range_vout` (testcontainers, runs in CI).
   - `services/asset/httpimpl/GetUTXOs_test.go::TestGetUTXOs/panic_in_goroutine_becomes_500_...` — mocks the repo to panic and asserts 500, locking the panic-isolation contract.

## Minor

- **413 mapping.** `io.ReadAll` now detects `echo.ErrStatusRequestEntityTooLarge` and surfaces 413 explicitly, so streaming clients without `Content-Length` get the documented status instead of a generic 500.
- **Counter only on success.** Acknowledged — mirrors `GetUTXO` and other handlers. Happy to add failure label-vector counters as a follow-up if you'd like; left out of this PR to keep parity with the rest of the asset surface.
- **Message-string behaviour change.** Confirmed no internal client greps the old strings.

## Nits

- `i := i` dropped.
- `writeUTXOsRecord` doc comment now uses `spendingVin`/`spendingTxID`, matching the PR description and `asset_reference.md`.

## Re @oskarszoon's items

- **LockTime divergence.** Pre-existing and non-trivial: SQL returns `coinbaseSpendingHeight`; Aerospike's stored `LockTime` bin is the tx's `nLockTime` (`stores/utxo/aerospike/create.go:714`) — semantically different. Documented as a known gap in `asset_reference.md`; reconciling the two stores is out of scope for this PR.
- **Anonymous + unrated endpoint.** Agreed this needs the auth/rate-limit work from #643. Will open a tracking issue referencing the `asset_httpHeavy` style gating once #643 lands.

## Re the github-actions bot's "error format mismatch"

The format string `"error getting utxo %s:%d"` with three args (`txHash, vout, err`) is the project's standard pattern, not a bug: `errors.New` strips a trailing `error` argument and attaches it as the wrapped error (`errors/errors.go:328-337`), then formats the remaining params. The wrapped error shows in the chain via `%w`-style unwrap, not via `%v` interpolation. Same pattern in `stores/utxo/aerospike/spend.go` etc.

**Comment by ordishs** [comment]
## Re-review — all blocking feedback addressed ✅

Re-reviewed at `830258f26` (fix commit `287eb7085`, *"prevent remote panic via out-of-range vout"*). Every point from the previous review is resolved, and the fix matches the recommended approach.

### 🔴 Critical DoS — fixed at the root
`stores/utxo/aerospike/get.go` now bounds-checks the batch offset before indexing and returns `NOT_FOUND` on overflow, mirroring the `ErrKeyNotFound` branch:
```go
idx := spend.Vout % uint32(s.utxoBatchSize)
if int(idx) >= len(utxos) {
    return &utxo.SpendResponse{Status: int(utxo.Status_NOT_FOUND)}, nil
}
b, ok := utxos[idx].([]byte)
```
Verified correct across single-batch, cross-batch, and exact-boundary vouts (e.g. 200-output tx: `vout=199`→valid, `vout=200`→NOT_FOUND via offset guard, `vout=300`→NOT_FOUND via missing record). SQL was already safe; both stores now have explicit coverage.

### 🟢 Defense-in-depth — added
`GetUTXOs.go` wraps each `errgroup` goroutine body in a `recover()` that logs `debug.Stack()` and converts any panic into a 500, so a future store-level panic degrades the request instead of taking down the process — exactly the isolation a public fan-out endpoint needs.

### 🟢 413 mapping — added and confirmed live
`io.ReadAll` now maps `echo.ErrStatusRequestEntityTooLarge` → 413. I checked that `teranode/errors.Is` delegates to stdlib `errors.Is` (`errors/errors.go:903`), so the foreign echo sentinel actually matches — not dead code. The documented 413 contract now holds for both `Content-Length` and chunked clients.

### 🟢 Test gap — closed at all three levels
- `stores/utxo/aerospike/aerospike_server_test.go` → `aerospike_get_spend_out_of_range_vout` (the real production store; would have panicked pre-fix).
- `stores/utxo/sql/sql_test.go` → `out-of-range vout returns NOT_FOUND, not error or panic`.
- `GetUTXOs_test.go` → `panic in goroutine becomes 500 instead of crashing the process` (forces a mock panic, asserts 500).

The `i := i` nit was also dropped (Go 1.26 loop semantics).

### Verification I ran
- `go test ./stores/utxo/sql/` → **PASS**, including all three `TestGetSpendNilUTXOHash` subcases (nil-skips-check, wrong-hash-fails, out-of-range→NOT_FOUND).
- The `httpimpl` and `repository`/aerospike tests I could **not** run in my environment — they transitively import the `gobdk` cgo module and my local Xcode license isn't accepted (cgo-on blocked, cgo-off excludes gobdk). Reviewed those by inspection; logic is sound. Worth a glance at CI to confirm the `-race` run is green on `services/asset/httpimpl/...` and the aerospike integration suite.

### Non-blocking
- The 413 read-path branch has no test (hard to exercise without a reader that lies about `Content-Length`) — optional.
- `prometheusAssetHTTPGetUTXOs` is still only incremented on success, so the bulk endpoint's failure rate is invisible — pre-existing pattern, fine to defer.

LGTM once CI is green. Nice, thorough turnaround.

**Review by oskarszoon — APPROVED** [review]
Approve at `830258f260`. DoS chain closed at both layers (defense-in-depth verified).

### The critical

`aerospike/get.go:192-197` bounds-check returns `Status_NOT_FOUND` before slice index — matches existing `ErrKeyNotFound` shape. `GetUTXOs.go:141-156` errgroup per-goroutine `defer recover()` is the first statement, named return propagates panic→500 through `g.Wait()`. Surgical fix + general belt.

Tests cover all three layers:
- `sql_test.go:396-413` — sqlitememory cross-store contract.
- `aerospike_server_test.go:226-251` — testcontainers, vout=99 on 2-output tx (panics pre-fix).
- `GetUTXOs_test.go:255-281` — handler-level panic→500.

All 673 tests `-race` clean.

### Prior items addressed

- LockTime divergence (SQL coinbase maturity height vs Aerospike `nLockTime`) documented in `asset_reference.md:790`.
- `asset_httpBodyLimit` cap + 413 mapping documented in `asset_reference.md:788-789`.
- UTXOHash nil-skip design noted in both store code comments.
- 413 mapping for `echo.ErrStatusRequestEntityTooLarge` cleanly separates 4xx from 5xx — better ops alerting.

### Follow-ups (not blocking)

- Operator-settings doc (`docs/references/settings/services/asset_settings.md`) doesn't list `asset_httpBodyLimit` yet — one-line add.
- Recover handler emits a full stack trace per panicking record. Closed paths are safe; if a future new panic source ever slips through, 116k records × ~5KB ≈ ~580MB log output per malicious request. Sample first stack only.
- Anonymous + unrated endpoint stays on the follow-up list pending #643.

**Review by ordishs — APPROVED** [review]
Approving — CI is green, including the `services/asset/httpimpl` race run and the aerospike out-of-range-vout regression test.

The critical remote-panic DoS (out-of-range vout → index-out-of-range in the aerospike store, surfacing in an errgroup goroutine outside echo's recover) is fixed at the root via the batch-offset bounds check in `aerospike/get.go`, with defense-in-depth recover in the fan-out goroutines, the 413 mapping, and regression tests at the aerospike, SQL, and handler layers. All earlier feedback addressed. Nice turnaround.

File: services/asset/httpimpl/swagger_routes.go [hunk 3]
```diff
@@ -374,6 +374,41 @@ package httpimpl
...
 //   404: errorResponse
 
+// swagger:route POST /api/v1/utxos utxo getUTXOsBulk
+// Bulk UTXO spend-status lookup (binary response).
+// Request body: concatenated 36-byte records, each [32 bytes txid][4 bytes vout little-endian].
+// Response body: concatenated fixed-length 48-byte records in input order, each
+// [8 bytes status LE][4 bytes lockTime LE][4 bytes spendingVin LE][32 bytes spendingTxID].
+// Unspent UTXOs have the trailing 36 bytes zero-filled; not-found records report
+// status utxo.Status_NOT_FOUND and the remaining bytes are zero. Body size is
+// capped by the asset_httpBodyLimit setting (default 4MB).
```
[the remark this answers is not in the document]
Comment (icellan) [reply services/asset/httpimpl/swagger_routes.go]: Fixed in 344d15966 — swagger now says `default 100MB` matching `asset_reference.md` and the post-merge global cap from PR #643. Thanks for catching this.

**Comment by icellan** [comment]
Addressed the latest bot/Sonar findings in 344d15966.

### Fixed
- **github-actions[bot] inline (swagger_routes.go:384)** — \`default 4MB\` → \`default 100MB\`. The merge from main brought in PR #643's global \`asset_httpBodyLimit\` (100MB); I'd missed updating the swagger annotation to match \`asset_reference.md\`.
- **SonarCloud CRITICAL — \`GetUTXOs\` cognitive complexity 31 (limit 15).** Extracted two helpers, no behaviour change:
  - \`readUTXOsBody(c)\` — body read + 413 mapping + length-multiple validation.
  - \`fetchOneUTXO(ctx, txHash, vout)\` — per-record store lookup with the ErrNotFound / nil-resp / wrapped-error semantics.
  The panic-recover defer stays inline in the goroutine wrapper because it needs the closure variables (\`txHash\`, \`vout\`, \`retErr\`).

### Flagging as false positive (please confirm)
- **SonarCloud 2× MAJOR — "Filename does not follow snake_case.go convention"** on \`GetUTXOs.go\` and \`GetUTXOs_test.go\`. Every \`Get*.go\` file under \`services/asset/httpimpl/\` (\`GetUTXO.go\`, \`GetTransaction.go\`, \`GetBlock.go\`, \`GetBlockHeader.go\`, ...) is CamelCase; my files match the established directory convention. Happy to rename if you'd prefer to align with Sonar (would need to rename the existing dozen+ files too for consistency).

All 10 \`TestGetUTXOs\` subtests pass under \`-race\`; \`make lint-new\` clean.
